README.md
## About
This is a small utility for providing a WebKit container targetted towards use on a embedded system via the framebuffer.

It can be considered a borderless/frameless/non-navigatable browser.

## Compilation

### Dependencies
```
apt-get install make g++ qt5-qmake qt5-default libqt5webkit5-dev gstreamer1.0-plugins-base gstreamer1.0-tools gstreamer1.0-x
```

### Compile

As a QT project compilation follows the standard QT pattern:

```
qmake # Generate Makefile
make  # Compile the project
```

### Debian Package

Native format package metadata is supplied to aid rolling your own package.

The standard build tools are supported:
```
debuild -uc -us
```



debian/changelog
qt5-webkit-container (20160123) UNRELEASED; urgency=medium

  * Initial project internal package.

 -- David Tulloh <debian-david@tulloh.id.au>  Mon, 25 Jan 2016 00:23:56 +1100


src/container.h
/* qt5-webkit-container - GPL2+ {{{
 * Copyright (C) 2016 Ian Firns
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 * }}} */

#ifndef CONTAINER_H
#define CONTAINER_H

#include <QtWebKitWidgets>

class Container : public QWebView {
    Q_OBJECT

  public:
    explicit Container(QWidget *parent = 0);

    void load(const QUrl &url);

    void setFixedSize(int w, int h);
    void setScrollBars(bool state);
    void setVerbosity(bool state);

    ~Container();

  private:
    QWidget *splash;
    QUrl    url;
    bool    verbose;

  private slots:
    void loadFinished(bool ok);
    void loadStarted();
    void retry();
};

#endif // CONTAINER_H


src/container.cpp
/* qt5-webkit-container - GPL2+ {{{
 * Copyright (C) 2016 Ian Firns
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 * }}} */

#include "container.h"
#include <stdio.h>

//#include <QScroller>

Container::Container(QWidget *parent) : QWebView(parent) {
  connect(this, SIGNAL(loadFinished(bool)), this, SLOT(loadFinished(bool)));
  connect(this, SIGNAL(loadStarted()), this, SLOT(loadStarted()));

  // enable kinetic scrolling
  //QScroller::grabGesture(this, QScroller::LeftMouseButtonGesture);

  // build splash screen
  splash = new QWidget(this);
  splash->setStyleSheet("background-color: #000");
  splash->setFixedSize(this->size());

  // no borders
  this->setWindowFlags(Qt::FramelessWindowHint);

  this->setScrollBars(false);
}

void Container::load(const QUrl &url) {
  this->url = url;
  QWebView::load(this->url);
}

void Container::setFixedSize(int w, int h) {
  if (verbose) {
    fprintf(stderr, "Setting container size: %d x %d\n", w, h);
  }

  QWebView::setFixedSize(w, h);

  // adjust splash appropriately
  splash->setFixedSize(w, h);
}

void Container::setScrollBars(bool state) {
  // show scrollbars as needed
  if (state) {
    this->page()->mainFrame()->setScrollBarPolicy(Qt::Vertical,   Qt::ScrollBarAsNeeded);
    this->page()->mainFrame()->setScrollBarPolicy(Qt::Horizontal, Qt::ScrollBarAsNeeded);
  }
  // disable scrollbars
  else {
    this->page()->mainFrame()->setScrollBarPolicy(Qt::Vertical,   Qt::ScrollBarAlwaysOff);
    this->page()->mainFrame()->setScrollBarPolicy(Qt::Horizontal, Qt::ScrollBarAlwaysOff);
  }
}

void Container::setVerbosity(bool state) {
  verbose = state;
}

// PRIVATE SLOTS
void Container::loadFinished(bool ok) {
  if (ok) {
    splash->hide();

    quint64 bytes_received = page()->bytesReceived();

    if (verbose) {
      fprintf(stderr, "Loaded (%llu bytes).\n", bytes_received);
    }
  }
  else {
    QTimer::singleShot(2000, this, SLOT(retry()));
  }
}

void Container::loadStarted() {
  if (verbose) {
    fprintf(stderr, "Loading: %s\n", (const char *)this->url.toString().toLatin1());
  }
}

void Container::retry() {
  if (verbose) {
    fprintf(stderr, "Retrying ...\n");
  }

  QWebView::load(this->url);
}

Container::~Container() {
  delete splash;
}


src/main.cpp
/* qt5-webkit-container - GPL2+ {{{
 * Copyright (C) 2016 Ian Firns
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 * }}} */

#include <iostream>

#include <QApplication>
#include <QCommandLineParser>
#include <QDebug>

#include "container.h"

#define DEFAULT_NAME     "qt5-webkit-container"
#define DEFAULT_VERSION  "UNRELEASED"

#define DEFAULT_URL      "http://localhost:3000"

#define DEFAULT_WIDTH    800
#define DEFAULT_HEIGHT   480


int main(int argc, char *argv[]) {
  QApplication a(argc, argv);

  a.setApplicationName(DEFAULT_NAME);
  a.setApplicationVersion(DEFAULT_VERSION);

  // process commandline
  QCommandLineParser parser;
  parser.setApplicationDescription(DEFAULT_NAME);
  parser.addHelpOption();
  parser.addVersionOption();

  QCommandLineOption verboseOption("verbose",
      QCoreApplication::translate("main", "Show verbose messaging."));

  parser.addOption(verboseOption);

  // add container size parsing
  QCommandLineOption widthOption("width",
      QCoreApplication::translate("main", "Set <width> of container."),
      QCoreApplication::translate("main", "width"));
  parser.addOption(widthOption);

  QCommandLineOption heightOption("height",
      QCoreApplication::translate("main", "Set <height> of container."),
      QCoreApplication::translate("main", "height"));
  parser.addOption(heightOption);

  // add cursor option parsing
  QCommandLineOption cursorOption(QStringList() << "c" << "no-cursor",
      QCoreApplication::translate("main", "Hide cursor."));
  parser.addOption(cursorOption);

  // add url option parsing
  QCommandLineOption urlOption(QStringList() << "u" << "url",
      QCoreApplication::translate("main", "Load <url> on startup."),
      QCoreApplication::translate("main", "url"));
  parser.addOption(urlOption);

  // process the actual command line arguments given by the user
  parser.process(a);

  // GPL requires that interactive programs inform the user of the licence
  std::cout << std::endl
		<< DEFAULT_NAME << ", Copyright (C) 2016 Ian Firns"                 << std::endl
		<< std::endl
		<< DEFAULT_NAME << " comes with ABSOLUTELY NO WARRANTY."           << std::endl
		<< "This is free software, and you are welcome to redistribute it" << std::endl
		<< "under certain conditions, see LICENCE file for details."       << std::endl
		<< std::endl;

  // process container size
  int width = parser.value(widthOption).toInt();
  int height = parser.value(heightOption).toInt();

  if (width <= 0) {
    width = DEFAULT_WIDTH;
  }

  if (height <= 0) {
    height = DEFAULT_HEIGHT;
  }

  // hide cursor as appropriate
  if (parser.isSet(cursorOption)) {
    QCursor cursor(Qt::BlankCursor);
    a.setOverrideCursor(cursor);
    a.changeOverrideCursor(cursor);
  }

  // process url
  QString url = parser.value(urlOption);
  if (url.length() == 0) {
    url = DEFAULT_URL;
  }

  // build container
  Container c;

  c.setVerbosity(parser.isSet(verboseOption));
  c.setFixedSize(width, height);
  c.load(QUrl(url));
  c.show();

  return a.exec();
}
